src/theory/sets/theory_sets_private.cpp

```diff
@@ -824,160 +552,176 @@ void TheorySetsPrivate::checkDownwardsClosure( std::vector< Node >& lemmas ) {
+                {
+                  Node x = itm2m.second[0];
+                  Node rr = d_equalityEngine.getRepresentative(term);
                   if( !isMember( x, rr ) ){
                     std::vector< Node > exp;
-                    exp.push_back( itm2m->second );
-                    addEqualityToExp( it2->second[1], itm2m->second[1], exp );
-                    Node k = getProxy( it2->second );
-                    Node fact = NodeManager::currentNM()->mkNode( kind::MEMBER, x, k );
-                    assertInference( fact, exp, lemmas, "upc2" );
-                    if( d_conflict ){
+                    exp.push_back(itm2m.second);
+                    d_state.addEqualityToExp(term[1], itm2m.second[1], exp);
+                    Node k = d_state.getProxy(term);
+                    Node fact = nm->mkNode(kind::MEMBER, x, k);
+                    d_im.assertInference(fact, exp, "upc2");
+                    if (d_state.isInConflict())
+                    {
                       return;
                     }
                   }
                 }
               }
-            }   
-          }      
+            }
+          }
         }
       }
     }
   }
-  if( !hasProcessed() ){
+  if (!d_im.hasProcessed())
+  {
     if( options::setsExt() ){
       //universal sets
       Trace("sets-debug") << "Check universe sets..." << std::endl;
       //all elements must be in universal set
-      for( std::map< Node, std::map< Node, Node > >::iterator it = d_pol_mems[0].begin(); it != d_pol_mems[0].end(); ++it ){
+      const std::vector<Node>& sec = d_state.getSetsEqClasses();
+      for (const Node& s : sec)
+      {
         //if equivalence class contains a variable
-        std::map< Node, Node >::iterator itv = d_var_set.find( it->first );
-        if( itv!=d_var_set.end() ){
+        Node v = d_state.getVariableSet(s);
+        if (!v.isNull())
+        {
           //the variable in the equivalence class
-          Node v = itv->second;
           std::map< TypeNode, Node > univ_set;
-          for( std::map< Node, Node >::iterator it2 = it->second.begin(); it2 != it->second.end(); ++it2 ){
-            Node e = it2->second[0];
+          const std::map<Node, Node>& smems = d_state.getMembers(s);
+          for (const std::pair<const Node, Node>& it2 : smems)
+          {
+            Node e = it2.second[0];
             TypeNode tn = NodeManager::currentNM()->mkSetType( e.getType() );
             Node u;
             std::map< TypeNode, Node >::iterator itu = univ_set.find( tn );
             if( itu==univ_set.end() ){
-              std::map< TypeNode, Node >::iterator itu = d_eqc_univset.find( tn );
+              Node ueqc = d_state.getUnivSetEqClass(tn);
               // if the universe does not yet exist, or is not in this equivalence class
-              if( itu==d_eqc_univset.end() || itu->second!=it->first ){
-                u = getUnivSet( tn );
+              if (itu->second != ueqc)
```

Comment:
When the execution reaches this line, the condition  `itu == d_eqc_univset.end() ` evaluates to `true`. Is it safe to dereference the iterator `it` in `itu->second != ueqc`? Is it possible to replace this condition with `ueqc.isNull()`?

Reply:
Yes, good catch.

Reply:
I actually meant to change this to `ueqc != it->first`.